lib/rust/ensogl/component/list-editor/src/lib.rs

```diff
@@ -0,0 +1,951 @@
+    pub fn new(cursor: &Cursor) -> Self {
+        let frp = Frp::new();
+        let root = display::object::Instance::new();
+        let layouted_elems = display::object::Instance::new();
+        let dragged_elems = display::object::Instance::new();
+        root.add_child(&layouted_elems);
+        root.add_child(&dragged_elems);
+        let model = default();
+        layouted_elems.use_auto_layout(); //.set_gap((GAP, GAP));
+        let add_elem_icon = Rectangle().build(|t| {
+            t.set_size(Vector2::new(20.0, 20.0))
+                .set_color(color::Rgba::new(0.0, 1.0, 0.0, 1.0))
+                .set_inset_border(2.0)
+                .set_border_color(color::Rgba::new(0.0, 0.0, 0.0, 0.5));
+        });
+
+        let remove_elem_icon = Rectangle().build(|t| {
+            t.set_size(Vector2::new(20.0, 20.0))
+                .set_color(color::Rgba::new(1.0, 0.0, 0.0, 1.0))
+                .set_inset_border(2.0)
+                .set_border_color(color::Rgba::new(0.0, 0.0, 0.0, 0.5));
+        });
+        add_elem_icon.set_y(-30.0);
+        root.add_child(&add_elem_icon);
+        remove_elem_icon.set_y(-30.0);
+        remove_elem_icon.set_x(30.0);
+        root.add_child(&remove_elem_icon);
+        Self { frp, root, layouted_elems, dragged_elems, model, add_elem_icon, remove_elem_icon }
+            .init(cursor)
+    }
+
+    fn init(self, cursor: &Cursor) -> Self {
+        let scene = scene();
+        let frp = &self.frp;
+        let network = self.frp.network();
+        let root = &self.root;
+        let layouted_elems = &self.layouted_elems;
+        let dragged_elems = &self.dragged_elems;
+        let model = &self.model;
+
+        let add_elem_icon_down = self.add_elem_icon.on_event::<mouse::Down>();
+        let remove_elem_icon_down = self.remove_elem_icon.on_event::<mouse::Down>();
+        let on_down = self.layouted_elems.on_event_capturing::<mouse::Down>();
+        let on_up = scene.on_event::<mouse::Up>();
+        let on_move = scene.on_event::<mouse::Move>();
+        frp::extend! { network
+
+            frp.private.output.request_new_item <+ add_elem_icon_down.map(f!([model] (_) {
+                let index = model.borrow_mut().items.len();
+                Response::new_gui_interaction(index)
+            }));
+
+            frp.remove <+ remove_elem_icon_down.constant(0);
+
+            push_item_index <= frp.push.map(f!([model, layouted_elems] (item)
+                item.upgrade().map(|t| model.borrow_mut().push_item(&layouted_elems, (*t).clone()))
+            ));
+            on_item_pushed <- frp.push.map2(&push_item_index, |item, ix| Response::new_api_interaction((*ix, item.clone())));
+            frp.private.output.on_new_item <+ on_item_pushed;
+
+            // Do not pass events to children, as we don't know whether we are about to drag
+            // them yet.
+            eval on_down ([] (event) event.stop_propagation());
```

Comment:
As we agreed, this is a temp file needed for testing list builder. It will be fixed in the next PR as stated on the top of the file.

Also, to explain it to readers - we will be blocking events when dragging elements and passing them otherwise.